### PR TITLE
Thermal target data

### PR DESCRIPTION
**Pull request by martamau**
This PR adds capability to generate a thermal target distribution for qcbm, i.e. Boltzmann distribution for a 1D chain of spins with Transverse Field Ising Model Hamiltonian and open boundary conditions. 

Note the tricky change in cost_function, that was needed for increasing FLIP performances. 
```
# circuit = ansatz.get_executable_circuit(parameters)
        circuit = ansatz._generate_circuit(parameters)
```
Not sure what is the best way to handle this but wanted to point it out.

### CODE REVIEW

File: src/python/zquantum/qcbm/ansatz.py [hunk 1]
```diff
@@ -254,4 +259,63 @@ def get_number_of_parameters_by_layer(self) -> np.ndarray:
...
+def save_qcbm_ansatz_set(qcbm_ansatz_set: List[QCBMAnsatz], filename: str) -> None:
+    """Save a set of qcbm_ansatz to a file.
+
+    Args:
+        qcbm_ansatz_set (list): a list ansatz to be saved
+        file (str): the name of the file
+    """
+    dictionary = {
+        "schema": ANSATZSET_SCHEMA,
+        "qcbm_ansatz": [ansatz.to_dict() for ansatz in qcbm_ansatz_set]
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/ansatz.py:289]: I think the key here should be `qcbm_ansatz_set`

File: src/python/zquantum/qcbm/ansatz.py [hunk 2]
```diff
@@ -254,4 +259,63 @@ def get_number_of_parameters_by_layer(self) -> np.ndarray:
...
+        data = json.load(file)
+
+    qcbm_ansatz_set = []
+    for item in data["qcbm_ansatz"]:
+        qcbm_ansatz_set.append(
+            QCBMAnsatz(
+                number_of_layers=item["number_of_layers"],
+                number_of_qubits=item["number_of_qubits"],
+                topology=item["topology"],
+            )
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/ansatz.py:318]: This certainly works, but it might make more sense to create a `from_dict` method for the `QCBMAnsatz` class. Then you could write this as `QCBMAnsatz.from_dict(item)`. But this is not necessary.

File: src/python/zquantum/qcbm/cost_function.py [hunk 3]
```diff
@@ -52,7 +52,8 @@ def cost_function(
             (float): cost function value for given parameters
             zquantum.core.bitstring_distribution.BitstringDistribution: distribution obtained
         """
-        circuit = ansatz.get_executable_circuit(parameters)
+        # circuit = ansatz.get_executable_circuit(parameters)
+        circuit = ansatz._generate_circuit(parameters)
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/cost_function.py:56]: Can you explain more why you chose to do this? If it's an efficiency improvement/avoiding some bottleneck, it would be better if we could instead figure out how to avoid the bottleneck in the first place as it's generally not good practice to call private methods from outside the class itself.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
+import bisect
+import typing
+import numpy as np
+from zquantum.core.bitstring_distribution import BitstringDistribution
+from zquantum.core.utils import dec2bin, bin2dec, convert_tuples_to_bitstrings
+
+
+def int2ising(number: int, length: int) -> typing.List[int]:
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:8]: I would suggest renaming this to something more descriptive such as: `convert_integer_to_ising_bitstring`

Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:8]: Also, do you think that this function would better fit in `z-quantum-core` or no? Seems like it could be used elsewhere beyond the QCBM

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        length: length of the Ising bitstring (i.e. positive number of spins in the Ising system)
+    Returns:
+        Ising bitstring representation (1D array of +/-1)."""
+
+    binary_bitstring = dec2bin(number, length)
+    ising_bitstring = [bit * 2 - 1 for bit in binary_bitstring]
+    return ising_bitstring
+
+
+def ising2int(ising_bitstring: typing.List[int]) -> int:
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:21]: Same as comments above.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        Integer number representation of the Ising bitstring."""
+
+    binary_bitstring = [int((bit + 1) / 2) for bit in ising_bitstring]
+    number = bin2dec(binary_bitstring)
+    return number
+
+
+def get_random_hamiltonian_parameters(n_spins: int) -> typing.Tuple:
+    """Generates random h, J, and where h and J are arrays of random coefficients sampled from a normal distribution with zero mean and sqrt(n_spins) sd.
+    For reproducibilty, fix random generator seed in the higher level from which this function is called.
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:35]: Asking because it seems like you've thought this through, but why fix the seed at higher levels as opposed to passing it in as a keyword argument?

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        ising_bitstring: 1D array of +/-1.
+    Returns:
+        Integer number representation of the Ising bitstring."""
+
+    binary_bitstring = [int((bit + 1) / 2) for bit in ising_bitstring]
+    number = bin2dec(binary_bitstring)
+    return number
+
+
+def get_random_hamiltonian_parameters(n_spins: int) -> typing.Tuple:
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:33]: This function name could also be more descriptive as it's not obvious what hamiltonian parameters are being randomly generated.

Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:33]: Also, the types within the output tuple should be added to the function header.

Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:33]: Lastly, if this function is only used in a certain place, it might be better to make it private.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+            two_body_couplings[i, j] = two_body_couplings[j, i] = np.random.normal(
+                0, np.sqrt(n_spins)
+            )
+    return external_fields, two_body_couplings
+
+
+def get_thermal_target_distribution_dict(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:58]: Noting the TODO (I think it can be removed)

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        hamiltonian_parameters: values of hamiltonian parameters, namely external fields and two body couplings.
+
+    Returns:
+       Thermal target distribution.
+       Number of positive spins in the spin state. 
+    """
+    partition_function = 0
+    external_fields, two_body_couplings = hamiltonian_parameters
+    beta = 1.0 / temperature
+    distribution = {}
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:75]: This function seems to be generating objects representing bitstring distributions. If that's the case, why not use the `BitstringDistribution` object from `zquantum.core.bitstring_distribution`?

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 8 above)
Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:58]: Also, should `hamiltonian_parameters` be a tuple instead of a list since it has fixed size/types?

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+            energy -= ising_bitstring[i] * external_fields[i]
+            for j in range(n_spins):
+                if j == i + 1:
+                    energy -= (
+                        ising_bitstring[i]
+                        * ising_bitstring[j]
+                        * two_body_couplings[i, j]
+                    )
+                else:
+                    continue
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:90]: I think you could replace this loop with the following:
```
energy -= ising_bitstring[i] * ising_bitstring[i+1] * two_body_couplings[i, i+1]
```

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 4 above)
Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:8]: Ah, I see the naming follows `dec2bin` and `bin2dec`. Pinging @max-radin and @mstechly for input on naming choice

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        boltzmann_factor = np.exp(energy * beta)
+        partition_function += boltzmann_factor
+
+        binary_bitstring = convert_tuples_to_bitstrings([dec2bin(spin, n_spins)])[-1]
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:94]: If only expecting one entry in the output list, I would suggest indexing by `[0]` instead of `[-1]` (this might very well just be me being picky though, so it's not necessary to change).

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+        distribution[binary_bitstring] = boltzmann_factor
+
+    normalized_distribution = {
+        key: value / partition_function for key, value in distribution.items()
+    }
+
+    return normalized_distribution
+
+
+def get_target_bitstring_distribution(
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:104]: Going off previous comment, I think this should function should be combined with `get_thermal_target_distribution_dict` as you shouldn't need separate functions.

Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:104]: Alternatively, you could make `get_thermal_target_distribution_dict` private, but I think it would make more sense to just build the object as a `BitstringDistribution` directly.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+        get_thermal_target_distribution_dict(
+            n_spins, temperature, hamiltonian_parameters
+        )
+    )
+
+
+def cumulate(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:128]: Noting the `TODO`, this time seems like the docstring still needs to be added :D

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 8 above)
Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:58]: I think the typing suggestion applies to other places in this PR too.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+) -> typing.List[float]:
+    """Generates array of cumulative probabilities from target_distribution.
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:130]: I don't totally understand this function. Is the idea to do something like the following:
`{"00": 0.2, "01": 0.15, "10": 0.6, "11": 0.05}` -> `[0.2, 0.35, 0.95, 1.0]`?

Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:130]: If so, how do you know that the keys in the distribution dictionary are arranged in the proper order?

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+       List of cumulative probabilities from the target distribution.
+    """
+    probabilities = get_thermal_target_distribution_dict(
+        n_spins, temperature, hamiltonian_parameters
+    )
+    cumulative = []
+    tot = 0.0
+    for i in range(len(probabilities.keys())):
+        tot += list(probabilities.values())[i]
+        cumulative.append(tot)
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:145]: I think this could be done with:
```
for probability in probabilities.values():
    tot += probability
    cumulative.append(tot)
```

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+       Thermal target Bistring Distribution.
+    """
+    return BitstringDistribution(
+        get_thermal_target_distribution_dict(
+            n_spins, temperature, hamiltonian_parameters
+        )
+    )
+
+
+def cumulate(
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:125]: I would suggest a more descriptive name such as `get_cumulative_probability_distribution`

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+    )
+    cumulative = []
+    tot = 0.0
+    for i in range(len(probabilities.keys())):
+        tot += list(probabilities.values())[i]
+        cumulative.append(tot)
+    return cumulative
+
+
+# Sample from the inverse cumulative partition function
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:149]: Comment should go in the docstring

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+def sample(
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:150]: More descriptive name is necessary since `sample` could collide with many other packages.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+    n_samples: int,
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:154]: noting `TODO`

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 16 above)
Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states.py:125]: I would also make this private unless you imagine it would be used elsewhere

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+    """
+    cumulative = cumulate(n_spins, temperature, hamiltonian_parameters)
+    samples = np.zeros((n_samples, n_spins))
+    y = np.zeros(n_samples)
+    for i in range(n_samples):
+        y[i] = np.random.uniform(low=0, high=1)
+    for i in range(len(y)):
+        idx = bisect.bisect_right(cumulative, y[i])
+        ising_bitstring = int2ising(idx, n_spins)
+        samples[i, :] = ising_bitstring
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:172]: Can you explain what is happening here? I'm having trouble following the logic.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+    return np.asarray(samples)
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:173]: I think it should already be a numpy array, so np.asarray is not necessary

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+
+
+
+def get_thermal_sampled_distribution(
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:177]: Same logic as stated with regards to `get_thermal_target_distribution_dict`. I think the code would be cleaner with only one function that returned a `BitstringDistribution` object.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+    n_samples: int,
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:181]: `TODO`

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+    """Generates thermal states sample distribution
+    Args:
+        n_samples: the number of samples from the original distribution
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+       histogram_samples: keys are binary string representations and values are corresponding probabilities.
+    """
+    sample_distribution_dict = {}
+    samples = sample(n_samples, n_spins, temperature, hamiltonian_parameters)
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:192]: Since I had trouble figuring out what `sample` does, it's hard to figure out what is going on in the following lines. I will wait to review this function until a later point.

File: src/python/zquantum/qcbm/target_thermal_states.py [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+        Sampled Bitstring Distribution.
+    """
+    probabilities = get_thermal_sampled_distribution(
+        n_samples, n_spins, temperature, hamiltonian_parameters
+    )[0]
+    spin_mapping = get_thermal_sampled_distribution(
+        n_samples, n_spins, temperature, hamiltonian_parameters
+    )[1]
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states.py:233]: This is taking two calls to `get_thermal_sampled_distribution` which seems unnecessary. You should be able to do:
```
probabilities, spin_mapping = get_thermal_sampled_distribution(n_samples, n_spins, temperature, hamiltonian_parameters)
```

File: src/python/zquantum/qcbm/target_thermal_states_test.py [hunk 26]
```diff
@@ -0,0 +1,246 @@
+import unittest
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states_test.py:1]: This file should be moved to `tests/zquantum/qcbm/`

Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states_test.py:1]: Also, we have moved to using `pytest` instead of `unittest`. If you'd like help rewriting the following tests, I can help :)

File: src/python/zquantum/qcbm/target_thermal_states_test.py [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+
+SEED = 14943
+
+
+class TestThermalTarget(unittest.TestCase):
+    def test_int2ising(self):
+        # Given
+        number = 4
+        length = 5
+        expected_ising_bitstring = [-1, -1, 1, -1, -1]
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states_test.py:27]: Once migrated to `pytest`, it would be best to use `pytest.mark.parametrize` so that more cases can be tested.

File: src/python/zquantum/qcbm/target_thermal_states_test.py [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+
+        # When
+        ising_bitstring = int2ising(number, length)
+
+        # Then
+        self.assertListEqual(ising_bitstring, expected_ising_bitstring)
+
+    def test_ising2int(self):
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states_test.py:35]: Same comment as above

File: src/python/zquantum/qcbm/target_thermal_states_test.py [hunk 29]
```diff
@@ -0,0 +1,246 @@
...
+        ising_bitstring = [-1, -1, 1, -1, -1]
+        expected_number = 4
+
+        # When
+        number = ising2int(ising_bitstring)
+
+        # Then
+        self.assertEqual(number, expected_number)
+
+    def test_ising2int2ising(self):
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states_test.py:46]: Same comment as above

File: src/python/zquantum/qcbm/target_thermal_states_test.py (hunk 27 above)
Comment (simonwa7) [reply src/python/zquantum/qcbm/target_thermal_states_test.py:27]: Additionally, I would suggest checking some edge cases like `number=0` or `length=-1`. It might also require having some tests that check the logic for "make sure this fails if the length is negative"

File: src/python/zquantum/qcbm/target_thermal_states_test.py [hunk 30]
```diff
@@ -0,0 +1,246 @@
...
+        expected_number = 14
+        length = 5
+
+        # When
+        number = ising2int(int2ising(expected_number, length))
+
+        # Then
+        self.assertEqual(number, expected_number)
+
+    def test_get_random_hamiltonian_parameters(self):
```
Comment (simonwa7) [inline src/python/zquantum/qcbm/target_thermal_states_test.py:57]: I'm going to refrain from commenting on the rest of the tests since I expect more changes to `src/python/zquantum/qcbm/target_thermal_state.py`

File: steps/generate_target_distribution_test.py [hunk 31]
```diff
@@ -0,0 +1,57 @@
+import os
```
Comment (simonwa7) [inline steps/generate_target_distribution_test.py:1]: I very much appreciate these tests, but we are moving away from testing step functions, so this file can be removed. Apologies for the confusion

**Review by simonwa7 — DISMISSED** [review]

File: src/python/zquantum/qcbm/cost_function.py (hunk 3 above)
Comment (martamau) [reply src/python/zquantum/qcbm/cost_function.py:56]: I shared a word doc on quantum software channel on slack at the beginning of May with all the data and explanation for this :) Will send you the link

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 6 above)
Comment (martamau) [reply src/python/zquantum/qcbm/target_thermal_states.py:35]: Because if you fix it in here and you generate several random hamiltonians, you will end up having the same exact one. Also in general the seed should be fixed once at the beginning of a script as it's supposed to be a global variable :)

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 4 above)
Comment (max-radin) [reply src/python/zquantum/qcbm/target_thermal_states.py:8]: I don't think there was any special reason for using the names `dec2bin` and `bin2dec`. As I recall the names came from some other package. So something more explicit would probably be good.

File: src/python/zquantum/qcbm/target_thermal_states_test.py [hunk 32]
```diff
@@ -3,9 +3,15 @@
 from zquantum.core.utils import dec2bin, convert_tuples_to_bitstrings
 from zquantum.core.bitstring_distribution import BitstringDistribution
 from zquantum.core.bitstring_distribution.distance_measures.mmd import compute_mmd
+<<<<<<< HEAD
```
Comment (mstechly) [inline src/python/zquantum/qcbm/target_thermal_states_test.py:6]: This needs to be fixed ;)

**Review by mstechly — CHANGES_REQUESTED** [review]

File: src/python/zquantum/qcbm/ansatz.py (hunk 1 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/ansatz.py]: Done!

File: src/python/zquantum/qcbm/ansatz.py (hunk 2 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/ansatz.py]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 4 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 5 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 7 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: I really don't know of a better name for this - even though I know this one is quite generic. 
Types = done. Private = done.

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 9 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:79]: There is a second function that turns this distribution into a BitstringDistribution object. Not sure if this is necessary.

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 10 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:94]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 12 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:108]: Agreed. Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 13 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 14 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:134]: It only works if you know that they are in the proper order, and for our case, we build the thermal state distribution to have the proper order, so it works. But for another random distribution, you would have to make sure it's in order first.

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 15 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:149]: Agreed. Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 16 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:129]: Done! Agreed!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 17 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:153]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 18 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:154]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 20 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: I believe its taking the cumulative distribution list, and finding the place where the random number between 0 and 1 should be sorted. So it's finding the index according to the distribution. I agree, it's a little weird.

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 22 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:180]: Agreed! Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 24 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:195]: Sample just generates random samples from the distribution

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 25 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py:233]: Taken out anyway

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 7 above)
Comment (martamau) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Change it to get_random_ising_hamiltonian_parameters and add details to docstring

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 8 above)
Comment (martamau) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 14 above)
Comment (martamau) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Check if it's sorted and if not we sort it, using https://github.com/zapatacomputing/z-quantum-core/blob/4680301d4b43d4de0888456291189b191ba1f04c/src/python/zquantum/core/utils.py#L645 + https://github.com/zapatacomputing/z-quantum-qcbm/blob/2572fcb386ea7b0542b7290d0f504ce88b668858/examples/plot_qcbm_opt_history.py#L45

File: src/python/zquantum/qcbm/target_thermal_states_test.py (hunk 32 above)
Comment (martamau) [reply src/python/zquantum/qcbm/target_thermal_states_test.py]: Done!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 7 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Input the name of the hamiltonian model - didn't write out the formula.

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 14 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Took out cumulate anyway as we switch to Manuel's function from z-quantum-core!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 20 above)
Comment (kaitlinmgili) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Took out sample as we use Manuel's function from z-quantum-core!

File: src/python/zquantum/qcbm/target_thermal_states.py (hunk 7 above)
Comment (martamau) [reply src/python/zquantum/qcbm/target_thermal_states.py]: Made a small correction (TFIM -> IM with non-zero external fields)

File: src/python/zquantum/qcbm/cost_function.py (hunk 3 above)
Comment (mstechly) [reply src/python/zquantum/qcbm/cost_function.py:57]: I would suggest adding a comment with TODO here explaining why we're doing this.
Having TODO will help us remember to actually change this once we have proper solution in place.

**Review by mstechly — APPROVED** [review]

File: src/python/zquantum/qcbm/cost_function.py (hunk 3 above)
Comment (mstechly) [reply src/python/zquantum/qcbm/cost_function.py:57]: Added it myself ;)
